single_kernel_mongo/managers/cluster.py

```
@@ -0,0 +1,351 @@
+            raise DeferrableFailedHookChecksError(
+                "mongos is integrated to a different CA than the config server. Please use the same CA for all cluster components."
+            )
+        if self.state.upgrade_in_progress:
+            raise DeferrableError(
+                "Processing client applications is not supported during an upgrade. The charm may be in a broken, unrecoverable state."
+            )
+
+    def relation_created(self) -> None:
+        """Just sets a status on relation created."""
+        logger.info("Integrating to config-server")
+        self.charm.status_manager.to_waiting("Connecting to config-server")
+
+    def on_database_created(self, username: str | None, password: str | None):
+        """Database created event.
+
+        Stores credentials in secrets and share it with clients.
+        """
+        if not username or not password:
+            raise WaitingForSecretsError
+        if self.state.upgrade_in_progress:
+            logger.warning(
+                "Processing client applications is not supported during an upgrade. The charm may be in a broken, unrecoverable state."
+            )
+            raise DeferrableFailedHookChecksError
+
+        if not self.charm.unit.is_leader():
+            return
+
+        logger.info("Database and user created for mongos application.")
+        self.state.secrets.set(AppPeerDataKeys.username.value, username, Scope.APP)
+        self.state.secrets.set(AppPeerDataKeys.password.value, password, Scope.APP)
+
+    def relation_changed(self) -> None:
+        """Start/restarts mongos with config server information."""
+        self.assert_pass_hook_checks()
+        key_file_contents = self.state.cluster.keyfile
+        config_server_db_uri = self.state.cluster.config_server_uri
+
+        if not key_file_contents or not config_server_db_uri:
+            raise WaitingForSecretsError("Waiting for keyfile or config server db uri")
```

do we really want to raise this error instead of one of the Deferrable errors?

---

Hum this is a bit more precise IMHO in case we want to keep track of the code in case of errors.
It could be a DeferrableError indeed, so I don't have a strong opinion here.

---

My thought was that most of the event handlers try except for DeferrableFailedHookChecksError and Non DeferrableFailedHookChecksError, so it might be best to be consistent in that approach.

---

No: They try/except on all errors.
Failed hook checks will raise `DeferrableFailedHookChecksError` or `NonDeferrableFailedHookChecksError`.
The actual processing of the event can raise other issues.
My idea is to keep in the most accurate errors outside of that so that we keep some context and you can see in the other events handlers/managers that we handle way more errors.
What could be done in a future would be to have a bit of inheritance however to have an easier error handling at the end.